test/edgehog_device_forwarder/termination_callbacks_test.exs

```diff
@@ -0,0 +1,33 @@
+# Copyright 2023 SECO Mind Srl
+# SPDX-License-Identifier: Apache-2.0
+
+defmodule EdgehogDeviceForwarder.TerminationCallbacksTest do
+  use ExUnit.Case
+  alias EdgehogDeviceForwarder.TerminationCallbacks
+
+  setup do
+    # Reset process monitoring
+    Supervisor.terminate_child(EdgehogDeviceForwarder.Supervisor, Monitor)
+    Supervisor.restart_child(EdgehogDeviceForwarder.Supervisor, Monitor)
+
+    process = spawn(fn -> wait_for_exit() end)
+    [process: process]
+  end
+
+  test "callback closure is executed once the process exits", %{process: process} do
+    message = :done
+    me = self()
+    send_message_to_self = fn -> send(me, message) end
+    TerminationCallbacks.add(process, send_message_to_self)
+
+    send(process, :exit)
+    assert_receive ^message
+  end
+
+  def wait_for_exit do
+    receive do
+      :exit -> :ok
+      _ -> wait_for_exit()
+    end
+  end
```

Comment:
Up what @rbino said, you don't need to recurse here. If you only put the `:exit -> :ok` branch in `receive`, `process` will block until `:exit` is received. Then, it will return `:ok` and exit. It seems that this is what you need